### PR TITLE
DEPR: Deprecate Index.set_value

### PR DESCRIPTION
**Pull request by topper-123**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Deprecates ``Index.set_value``. This is a very little used and confusing method IMO.

For a given index ``idx``, array ``arr``,  value ``idx_val`` in ``idx`` and a new value of ``val``, ``idx.set_value(arr, idx_val, val)``  is equivalent to ``arr[idx.get_loc(idx_val) = val``, which is more standard and should be used instead.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
There has been some discussion before about `Index.get_value`: https://github.com/pandas-dev/pandas/issues/19728, where the thought was that this might be relied upon.  
But in contrast to get_value, set_value is apparently something we don't use internally ourselves?

cc @shoyer

**Comment by topper-123** [comment]
Yeah that's right, ``Index.set_value`` is not used by pandas itself.

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -116,7 +116,9 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
 
--
+- :meth:`Index.set_values` has been deprecated. For a given index ``idx``, array ``arr``,
+  value in ``idx`` of ``idx_val`` and a new value of ``val``, ``idx.set_value(arr, idx_val, val)``
+  is equivalent to ``arr[idx.get_loc(idx_val) = val``, which should be used instead (:issue:`28621`).
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
  is equivalent to ``arr[idx.get_loc(idx_val) = val]``, which should be used instead (:issue:`28621`).
```

missing end bracket

**Review by WillAyd — COMMENTED** [review]
I think this makes sense so OK by me

File: doc/source/whatsnew/v1.0.0.rst [hunk 2]
```diff
@@ -116,7 +116,9 @@ Other API changes
 Deprecations
 ~~~~~~~~~~~~
 
--
+- :meth:`Index.set_values` has been deprecated. For a given index ``idx``, array ``arr``,
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: does this actually render? you should remove set_values from api.rst

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm ping on green

**Comment by topper-123** [comment]
I've updated the PR according to comments.

In addition, I've added ``"set_value"`` to ``Index._deprecations``.

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**